Allow user to specify own canvas context when rendering in headless mode

pixelspark opened the pull request:
Implements https://github.com/vega/vega/issues/2547

This allows a user to specify a pre-existing context to draw into (instead of Vega creating one). This can be used to render multiple charts to a single PDF document without quality loss. The example below renders two charts to the same PDF:

````js
const Vega = require("./packages/vega");
const fs = require('fs');
const { createCanvas, loadImage } = require('canvas');

// Try headless render
const spec = require("./spec.json");

async function main() {
	const view = new Vega.View(Vega.parse(spec), {
		logger: Vega.logger(Vega.Warn, 'error'),
		renderer: "none"
	}).width(640).height(480);

	const canvas = createCanvas(640, 960, "pdf");
	const ctx = canvas.getContext('2d');

	await view.toCanvas(4.0, {
		context: ctx
	});

	ctx.translate(0, 460);
	await view.toCanvas(4.0, {
		context: ctx
	});

	const out = fs.createWriteStream(__dirname + '/test.pdf');
	const stream = canvas.createPDFStream();
	stream.pipe(out);
	out.on('finish', () =>  console.log('The PDF file was created.'));
}

main();
````

jheer commented on packages/vega-scenegraph/src/CanvasRenderer.js:
```diff
@@ -31,9 +36,16 @@ prototype.initialize = function(el, width, height, origin, scaleFactor, options)
   base.resize.call(this, width, height, origin, scaleFactor);
-  resize(this._canvas, this._width, this._height,
-    this._origin, this._scale, this._options && this._options.context);
+  if (this._canvas) {
```
It looks like resize will never occur with a passed-in context, as `this._canvas` will be undefined?

pixelspark replied:
Oops, there should indeed not be an 'if' here!

jheer commented on packages/vega-scenegraph/src/CanvasRenderer.js:
```diff
@@ -31,9 +36,16 @@ prototype.initialize = function(el, width, height, origin, scaleFactor, options)
+    throw new Error('cannot resize canvas twice');
+  }
+
+  this._resized = true;
```
I'd prefer to have the logic prevent this from happening, alongside any relevant test cases, rather than add a new member variable and check an assertion at runtime.

pixelspark replied:
OK

jheer commented on packages/vega-scenegraph/src/util/canvas/resize.js:
```diff
@@ -4,21 +4,23 @@ function devicePixelRatio() {
+    canvas.height = height * ratio;
+  }
 
   for (const key in opt) {
```
With a passed-in context, I don't think we want to apply all the context-specific operations in the `resize` utility. For example, here we may overwrite desired properties (set externally) of the node-canvas context. That said, the bigger concern is that the transform matrix gets completely overwritten later in the method via `setTransform`!

pixelspark replied:
Applying setTransform to `context` is fine as long as the user calls `save` (and later `restore`) on it, right?

pixelspark commented:
OK, so I have now moved the resizing logic (translate + scale to specified scale factor) of the user-supplied context back to CanvasRenderer. The `setTransform` is replaced with calls to `translate` and `scale` as to not overwrite the user's transform.

Not sure how we should prevent calling resize() multiple times as it is not prevented for the regular case either. With regular use it is not called more than once and produces the correct results. I removed the runtime check.

pixelspark commented:
Just noticed that `toCanvas` already accepted `{context: ...}` as an option to set various context options. I now changed the option name to provide a custom context to `intoContext`. TypeScript typings have been changed accordingly, which ensures that you either specify `intoContext` *or* `context`/`type`, not both at the same time. (I considered re-using `context` and checking whether a context or an options object was supplied using `instanceof`, but that is tricky as this should both work in the browser as well as on the server).

@jheer please let me know if you have any further suggestions!

jheer commented:
Incorporated into PR #2575
